### PR TITLE
Fix cudnn Dropout reproducibility

### PR DESCRIPTION
**Pull request by roywei**
Fix https://github.com/apache/incubator-mxnet/issues/15662
This will replace #16532 with an alternative solution.
Please refer to the discussion in  https://github.com/apache/incubator-mxnet/pull/16532#discussion_r339268855

1. Added `GetSeed()` in GPU random resource, it's similar to CPU version.
2. During cudnn dropout, check whether random seed has been changed and reset cudnn dropout descriptor's seed if random seed changed.

Benefit:
1. This avoided generating a random int (tensor on gpu) on GPU, copy to CPU and and pass it to `cudnnDropoutGetStatesSize`.
2. get the seed from gpu random resource is fast, we can afford to do it every forward.

Drawback:
Will be affected by https://github.com/apache/incubator-mxnet/issues/7410, by default mxnet's random seed is fixed.

### CODE REVIEW

**Comment by roywei** [comment]
@DickJC123 @ptrendx could you help take a look again? Thanks

File: 3rdparty/mshadow/mshadow/random.h [hunk 1]
```diff
@@ -404,8 +404,16 @@ class Random<gpu, DType> {
...
     curandStatus_t status;
     status = curandSetPseudoRandomGeneratorSeed(gen_, static_cast<uint64_t>(seed));
+    this->rseed_ = static_cast<unsigned>(seed);
     CHECK_EQ(status, CURAND_STATUS_SUCCESS) << "Set CURAND seed failed.";
   }
+  /*!
+    * \brief get random seed used in random generator
+    * \return seed in unsigned
+    */
+  inline unsigned GetSeed() const {
```
Comment (apeforest) [inline 3rdparty/mshadow/mshadow/random.h]: Why can't we just use uint64_t as the data type for rseed_?

File: 3rdparty/mshadow/mshadow/random.h [hunk 2]
```diff
@@ -466,6 +474,7 @@ class Random<gpu, DType> {
   uniform(Shape<dim> shape);
 
  private:
+  unsigned rseed_;
```
Comment (apeforest) [inline 3rdparty/mshadow/mshadow/random.h]: use uint64_t?

File: src/operator/nn/dropout-inl.h [hunk 3]
```diff
@@ -255,8 +255,13 @@ class DropoutOp {
       Stream<xpu> *s = ctx.get_stream<xpu>();
 
       // set dropout state.
-      ctx.requested[0].get_cudnn_dropout_desc(&dropout_desc_, s, 1.0f - this->pkeep_, seed_);
-
+      Random<xpu, unsigned> *prnd = ctx.requested[1].get_random<xpu, unsigned>(s);
+      uint64_t rng_seed = (uint64_t) prnd->GetSeed();
```
Comment (apeforest) [inline src/operator/nn/dropout-inl.h]: If GetSeed() returns uint64_t you don't need the cast here

File: src/operator/rnn.cc [hunk 4]
```diff
@@ -180,6 +180,7 @@ static std::vector<ResourceRequest> RNNResourceEx(const NodeAttrs& attrs, const
     const RNNParam& param = nnvm::get<RNNParam>(attrs.parsed);
     if (param.p != 0 && 1.0f - param.p > 0) {
       request.emplace_back(ResourceRequest::kCuDNNDropoutDesc);
+        request.emplace_back(ResourceRequest::kRandom);
```
Comment (apeforest) [inline src/operator/rnn.cc]: nit: indent here not consistent.

**Comment by karan6181** [comment]
@DickJC123 @ptrendx Can you please review this PR once again? Thanks!

**Comment by apeforest** [comment]
@roywei Could you please rebase and update this PR? Thanks!

**Comment by ChaiBapchya** [comment]
@roywei with CI issue fixed. let's rebase to include the fix. Thanks!

**Comment by access2rohit** [comment]
restarted CI jobs. @roywei can you get your PR merged once they pass CI

**Comment by ChaiBapchya** [comment]
@access2rohit without rebasing, the fix #17962 won't be included [for windows-gpu]

**Comment by roywei** [comment]
@ChaiBapchya @access2rohit  Rebased! Thanks!

File: src/operator/rnn-inl.h [hunk 5]
```diff
@@ -1495,7 +1500,7 @@ class RNNOp {
   cudnnRNNInputMode_t input_mode_;
   cudnnDropoutDescriptor_t dropout_desc_;
   Storage::Handle reserve_space_;
-  uint64_t seed_ = 17 + rand() % 4096;  // NOLINT(runtime/threadsafe_fn)
```
Comment (leezu) [inline src/operator/rnn-inl.h:1498]: This `seed_` is used for `rand_r` in the forward implementations. That's very bad practice, not portable and should be fixed.
As this PR already passes CI, it's fine to fix it in a follow-up PR. In fac, https://github.com/apache/incubator-mxnet/pull/17984 is blocked by the usage of `rand_r` so I may fix it there.

More info: https://channel9.msdn.com/Events/GoingNative/2013/rand-Considered-Harmful

File: src/operator/nn/dropout-inl.h [hunk 6]
```diff
@@ -255,8 +255,13 @@ class DropoutOp {
...
       // set dropout state.
-      ctx.requested[0].get_cudnn_dropout_desc(&dropout_desc_, s, 1.0f - this->pkeep_, seed_);
-
+      Random<xpu, unsigned> *prnd = ctx.requested[1].get_random<xpu, unsigned>(s);
+      uint64_t rng_seed = prnd->GetSeed();
+      // reset dropout descriptor if rng seed changed.
+      bool reset = seed_ != rng_seed;
+      seed_ = rng_seed;
+      ctx.requested[0].get_cudnn_dropout_desc(&dropout_desc_, s, 1.0f - this->pkeep_,
+          seed_, reset);
```
Comment (szha) [inline src/operator/nn/dropout-inl.h:264]: @roywei there are more than 1 random number generators in the resources (4 by default) and they could have different seeds. The result of rotating random number generator seeds here is that cudnn dropout state is reinitialized very often. @sxjscience observed that there's significant performance impact because of this.

Comment (rondogency) [reply src/operator/nn/dropout-inl.h:264]: I believe @roywei is already working on this issue, since this regression has been caught by our benchmark last week

Comment (roywei) [reply src/operator/nn/dropout-inl.h:264]: I spent some time looking into it and actually the problem is dropout's forward is entering the [re-init cudnn dropout desc logic](https://github.com/apache/incubator-mxnet/pull/17547/files#diff-cc7bb408eba92cdd6fad0e590b76fabeL438) every time during forward even without my change. This is true for nd/np and gluon, false for symbol dropout. So originally it was already reinitializing in every forward. The reason it does not cause any performance regression is because it's always using the `seed_` defined in `uint64_t seed_ = 17 + rand() % 4096; ` and never changed, it won't listen to MXNet's PRNG (which is the problem this PR is trying to fix). So if the seed didn't change, event if you re-init cudnn dropout descriptor, it won't take any time. My PR changed the seed, so it was actually re-init every forward, thus the regression.

However, it works fine under symbol case. I guess is somehow for symbolic, every forward is using the same Dropout node and the state check took effect the didn't go into re-init logic. For imperative case the state check is always empty and went into the re-init logic.

Compare the following two code, one is gluon and one is symbol, if I print some log during reinitialization in the original code without my change [here](https://github.com/apache/incubator-mxnet/pull/17547/files#diff-cc7bb408eba92cdd6fad0e590b76fabeL438). It will print out every forward in ND/NP/Gluon, not the case in Symbol.
```
import mxnet as mx
data = mx.nd.ones((10, 200, 300, 500), ctx=mx.gpu(0))
dropout = mx.gluon.nn.Dropout(0.5)
# with or without hybridize is the same result
dropout.hybridize()
with mx.autograd.record():
        result1 = dropout(data)
        result2 = dropout(result1)
```
print 2 times
```
re-init dropout desc
re-init dropout desc
```

Symbol:
```
import mxnet as mx
data = mx.nd.ones((10, 200, 300, 500), ctx=mx.gpu(0))
net = mx.sym.Variable("data")
net = mx.sym.Dropout(data=net, p=0.5, cudnn_off=False)
exe = net.simple_bind(mx.gpu(0), data=data.shape)
result1 = exe.forward(is_train=True, data=data)
result2 = exe.forward(is_train=True, data=result1[0])
```

print 1 time
```
re-init dropout desc
```

Given this situation, I don't have a good solution as checking the state handle size does not work in imperative mode, checking PRNG seed will also not work. I will revert this PR for now as it's causing regressions for models using dropout.

cc @szha @sxjscience @apeforest

Comment (szha) [reply src/operator/nn/dropout-inl.h:264]: @roywei the links in your comment don't seem to point to what they refer to so I'm a bit confused. Where did you insert the 're-init dropout desc' comment?
